sdk/core/azure-core/test/ut/url.cpp

```diff
@@ -0,0 +1,226 @@
+        req_with_query.GetUrl().GetAbsoluteUrl(),
+        "http://test.com?query=value");
+
+    // retry query params testing
+    req_with_query.StartTry();
+    // same query parameter should override previous
+    EXPECT_NO_THROW(req_with_query.GetUrl().AppendQueryParameter("query", "retryValue"));
+
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req_with_query.GetUrl().GetAbsoluteUrl(),
+        "http://test.com?query=retryValue");
+  }
+
+  TEST(URL, query_parameter_encode_decode)
+  {
+    Http::HttpMethod httpMethod = Http::HttpMethod::Put;
+    Http::Url url("http://test.com");
+    EXPECT_NO_THROW(url.AppendQueryParameter("query", Http::Url::Encode("va=lue")));
+
+    // Default encoder from URL won't encode an equal symbol
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va%3Dlue");
+
+    // Provide symbol to do not encode
+    EXPECT_NO_THROW(url.AppendQueryParameter("query", Http::Url::Encode("va=lue", "=")));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va=lue");
+
+    // Provide more than one extra symbol to be encoded
+    EXPECT_NO_THROW(url.AppendQueryParameter("query", Http::Url::Encode("va=l u?e", " ?")));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va%3Dl u?e");
+
+    // default, encode it all
+    EXPECT_NO_THROW(url.AppendQueryParameter("query", Http::Url::Encode("va=l u?e")));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        url.GetAbsoluteUrl(),
+        "http://test.com?query=va%3Dl%20u%3Fe");
+  }
+
+  TEST(URL, add_path)
+  {
+    Http::HttpMethod httpMethod = Http::HttpMethod::Post;
+    Http::Url url("http://test.com");
+    Http::Request req(httpMethod, url);
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendQueryParameter("query", "value"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path?query=value");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path2"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path/path2?query=value");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path3"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path/path2/path3?query=value");
+  }
+
+  TEST(URL, getPort)
```

Comment:
Add a GetPort test without setting the port, to verify we return -1 by default. I wonder if we should throw in such cases instead. We probably don't want invalid ports being returned.

Reply:
Think I already added a test for that.
I wouldn't like to throw because is more expensive than checking for -1.
This `GetPort()` api is usually expected for transport adapters only. They need to know if a specific port was set for the url or not.
They will check for != -1 without adding a try/catch for that

Reply:
Good point.